_posts/2018-02-27-optimizing-grpc-part-1.md

```diff
@@ -0,0 +1,395 @@
+RPC couldn't start until the last one completed.  While slow, it also prevented us from creating 
+tons of RPCs that we didn't have memory for.  We need to account for this in the listenable 
+future version.
+
+To solve this, we can apply a self-imposed limit on the number of active RPCs.  Before starting a 
+new RPC, we will try to acquire a permit.  If we get one, the RPC can start.  If not, we will wait
+until one is available.  When an RPC completes (either in success or failure), we return the 
+permit.  To [accomplish](https://github.com/carl-mastrangelo/kvstore/blob/02-future-client/src/main/java/io/grpc/examples/KvClient.java#L94)
+this, we will using a `Semaphore`:
+
+```java
+  private final Semaphore limiter = new Semaphore(100);
+
+  private void doCreate(KeyValueServiceFutureStub stub, AtomicReference<Throwable> error)
+      throws InterruptedException {
+    limiter.acquire();
+    ByteString key = createRandomKey();
+    ListenableFuture<CreateResponse> res = stub.create(
+        CreateRequest.newBuilder()
+            .setKey(key)
+            .setValue(randomBytes(MEAN_VALUE_SIZE))
+            .build());
+    res.addListener(() ->  {
+      rpcCount.incrementAndGet();
+      limiter.release();
+    }, MoreExecutors.directExecutor());
+    /* ... */
+  }
+```
+
+Now the code runs successfully, and doesn't run out of memory.
+
+### Results
+
+Building and running the code again looks a lot better:
+
+```
+$ ./gradlew installDist
+$ time ./build/install/kvstore/bin/kvstore
+Feb 26, 2018 2:40:47 PM io.grpc.examples.KvRunner runClient
+INFO: Starting
+Feb 26, 2018 2:41:47 PM io.grpc.examples.KvRunner runClient
+INFO: Did 24.283 RPCs/s
+
+real	1m0.923s
+user	0m12.772s
+sys	0m1.572s
+```
+
+Our code does **46%** more RPCs per second than previously.  We can also see that we used about 20% 
+more CPU than previously.  As we can see our hypothesis turned out to be correct and the fix 
+worked.  All this happened without making any changes to the server.  Also, we were able to 
+measure without using any special profilers or tracers.
+
+Do the numbers make sense?  We expect to issue mutation (create, update, and delete) RPCs each 
+about with 1/4 probability.  Reads are also issue 1/4 of the time, but don't take as long.  The
+mean RPC time should be about the weighted average RPC time:
+
+```
+  .25 * 50ms (create)
+  .25 * 10ms (retrieve)
+  .25 * 50ms (update)
+ +.25 * 50ms (delete)
+------------
+        40ms
+```
+
+At 40ms on average per RPC, we would expect the number of RPCs per second to be:
+
+25 queries = 1000ms / (40 ms / query)
+
+That's approximately what we see with the new code.  The server is still serially handling 
+requests, so it seems like we have more work to do in the future.  But for now, our optimizations
+seem to have worked.
+
+
+
+## Conclusion
+
+There are a lot of opportunities to optimize your gRPC code to be faster.  To take advantage of 
```

Comment:
There are a lot of opportunities to optimize your gRPC code. To take..

Reply:
Done.